File: lib/std/core/types/player.lua
require 'std.core.util'
local hook = require 'std.core.hook'

player = player or {
    __playerStorage = {},
    __localPlayer = nil
}

local playerStorage = player.__playerStorage
local localPlayer = player.__localPlayer
local playerMeta = util.meta.new("player")

local function wrapPlayer(p)
    local id = GetPlayerId(p)
    if playerStorage[id] ~= nil then
        return playerStorage[id]
    end

    local t = {
        __id = GetPlayerId(p),
        __h = p
    }

    setmetatable(t, playerMeta)
    return t
end

_G.__WrapPlayer = wrapPlayer

function player.getLocal()
    return localPlayer
end

function player.get(id)
    ASSERT_ARG_TYPE(1, "id", "number")

    return playerStorage[id]
end

function player.getAll()
    local t = {}

    for i=0, bj_MAX_PLAYER_SLOTS - 1 do
        table.insert(t, playerStorage[i])
    end

    return t
end

function playerMeta.get:id()
    return self.__id
end

function playerMeta.get:name()
    return GetPlayerName(self.__h)
end

function playerMeta.set:name(name)
    ASSERT_ARG_TYPE(1, "name", "string")

    SetPlayerName(self.__h, name)
end

hook.add("init", function()
    for i=0, bj_MAX_PLAYER_SLOTS - 1 do
        player.__playerStorage[i] = wrapPlayer(Player(i))
    end

    player.__localPlayer = player.__playerStorage[GetPlayerId(GetLocalPlayer())]
end)

return player

File: ts/src/tsstd/core/types/player.ts
import Handle from "./handle"

let playerNative = _G.Player

class Player extends Handle<jplayer> {
    protected constructor(handle: jplayer, ctor: typeof Player) {
        super(handle, ctor as typeof Handle)
        print(`constructed Player for ${GetPlayerId(handle)}:${GetHandleId(handle)}`)
    }

    static local: Player
    static all: Player[]

    private static _init = util.contextFn("player/all", () => {
        Player.init()
        let storage: Player[] = []

        for (let i = 0; i < 4; i++) {
            storage[i] = Player.of(playerNative(i))
        }

        Player.all = storage
        Player.local = Player.of(GetLocalPlayer())
    })

    get id() {
        return GetPlayerId(this.handle)
    }

    get name() {
        return GetPlayerName(this.handle)
    }

    set name(v: string) {
        SetPlayerName(this.handle, v)
    }

    get isLocal() {
        return this == Player.local
    }

    get isHuman() {
        return GetPlayerController(this.handle) == MAP_CONTROL_USER
    }

    get isPlaying() {
        return GetPlayerSlotState(this.handle) == PLAYER_SLOT_STATE_PLAYING
    }

    destroy() {
        throw new Error("Player instances are not supposed to be destroyed.")
    }

    static byId(id: number): Player {
        return Player.all[id]!
    }

    static of(player: jplayer): Player {
        return Player.getForHandle(player)
    }
}

Player.init()

export default Player
